compiler/qsc_formatter/src/formatter.rs

```diff
@@ -0,0 +1,249 @@
+
+    // if the right is a close brace, the indent level should be one less
+    let indent_level = if let Syntax(Close(Delim::Brace)) = right.kind {
+        indent_level.saturating_sub(1)
+    } else {
+        indent_level
+    };
+
+    use qsc_frontend::keyword::Keyword;
+    use ConcreteTokenKind::*;
+    use TokenKind::*;
+    match (&left.kind, &right.kind) {
+        (Syntax(Open(l)), Syntax(Close(r))) if l == r => {
+            rule_no_space(left, whitespace, right, &mut edits);
+        }
+        (Comment | Syntax(DocComment), _) => {
+            rule_trim_comments(left, &mut edits, code);
+            rule_indentation(left, whitespace, right, &mut edits, indent_level);
+        }
+        (Syntax(Semi), _) => match &right.kind {
+            Comment => {
+                if whitespace.contains('\n') {
+                    rule_indentation(left, whitespace, right, &mut edits, indent_level);
+                }
+            }
+            _ => {
+                rule_indentation(left, whitespace, right, &mut edits, indent_level);
+            }
+        },
+        (_, Syntax(Close(Delim::Brace))) => {
+            rule_indentation(left, whitespace, right, &mut edits, indent_level);
+        }
+        (_, Syntax(Keyword(Keyword::Operation)))
+        | (_, Syntax(Keyword(Keyword::Function)))
+        | (_, Syntax(Keyword(Keyword::Newtype)))
+        | (_, Syntax(Keyword(Keyword::Namespace))) => {
+            rule_indentation(left, whitespace, right, &mut edits, indent_level);
+        }
+        (Syntax(Open(Delim::Brace)), _) => {
+            rule_indentation(left, whitespace, right, &mut edits, indent_level);
+        }
+        (Syntax(Close(Delim::Brace)), Syntax(Semi)) => {
+            rule_no_space(left, whitespace, right, &mut edits);
+        }
+        (Syntax(Close(Delim::Brace)), _) => {
+            rule_indentation(left, whitespace, right, &mut edits, indent_level);
+        }
+        (Syntax(cooked_left), Syntax(cooked_right)) => match (cooked_left, cooked_right) {
+            (Ident, Ident) | (Keyword(_), Ident) | (_, Colon) | (Colon, _) | (Comma, _) => {
+                rule_single_space(left, whitespace, right, &mut edits);
+            }
+            (Ident, Open(Delim::Paren))
+            | (Ident, Open(Delim::Bracket))
+            | (Ident, Comma)
+            | (Open(_), _)
+            | (_, Close(_)) => {
+                rule_no_space(left, whitespace, right, &mut edits);
+            }
+            _ => {}
+        },
+        _ => {}
+    }
+
+    println!(
+        "edits for `{}` : {edits:?}",
+        &code[left.span.lo as usize..right.span.hi as usize]
+    );
+    edits
+}
+
+fn rule_no_space(
+    left: &ConcreteToken,
+    whitespace: &str,
+    right: &ConcreteToken,
+    edits: &mut Vec<Edit>,
+) {
+    if !whitespace.is_empty() {
+        edits.push(Edit::new(left.span.hi, right.span.lo, ""));
+    }
+}
+
+fn rule_single_space(
```

Comment:
What does `rule_single_space` mean?

Reply:
Similar to the above comment, I've renamed the function. It is a function that is meant to ensure that the given whitespace has only one space in it, and if that isn't the case, it will generate an edit to make it so.